packages/microsoft-security-adapter/src/definitions/references/references.ts

```diff
@@ -35,6 +39,18 @@ export const REFERENCES: definitions.ApiDefinitions<Options>['references'] = {
       lookup: referenceUtils.basicLookUp,
       lookupIndexName: 'appId',
     },
+    servicePrincipalAppId: {
+      serialize: ({ ref }) => {
+        const { appId } = ref.value.value
+        if (isReferenceExpression(appId)) {
```

Comment:
nit: I would verify the type 
eg 
```suggestion
        if (isReferenceExpression(appId) && ref.value.elemID.typeName === 'ServicePrincipal') {
```

same in line 46, otherwise log error and throw

Reply:
@shir-reifenberg I think we should only log and throw an error if it is a reference, but not a reference to an application. This is because we may have Service Principals where this field was not resolved as a reference if the application does not exist—for example, for built-in Microsoft applications.

Reply:
Also, it looks like the serializer is running both on deploy and fetch (to validate the reference). We discussed this f2f and decided not to throw an error, but to only log it.